storage/transfermanager/integration_test.go

```diff
@@ -456,6 +457,288 @@ func TestIntegration_DownloadDirectoryError(t *testing.T) {
 	})
 }
 
+// TestIntegration_DownloadDirectorySkippedFiles tests that an error is returned if any
+// object download escapes target directory
+func TestIntegration_DownloadDirectoryWithSkippedFiles(t *testing.T) {
+	multiTransportTest(context.Background(), t, func(t *testing.T, ctx context.Context, c *storage.Client, tb downloadTestBucket) {
+		localDir := t.TempDir()
+		callbackMu := sync.Mutex{}
+		numCallbacks := 0
+		numErrors := 0
+		illegalObjs := []string{
+			"../objA",
+			"../objB",
+		}
+
+		b := c.Bucket(tb.bucket)
+		for _, obj := range illegalObjs {
+			size := randomInt64(minObjectSize, maxObjectSize)
+			_, err := generateFileInGCS(ctx, b.Object(obj), size)
+			if err != nil {
+				t.Errorf("could not create file in GCS: %v for obj %s", err, obj)
+			}
+		}
+		t.Cleanup(func() {
+			for _, obj := range illegalObjs {
+				if err := b.Object(obj).Delete(ctx); err != nil {
+					t.Errorf("failed to delete object in cleanup for %s: %v", obj, err)
+				}
+			}
+		})
+		d, err := NewDownloader(c, WithWorkers(8), WithPartSize(maxObjectSize/2))
+		if err != nil {
+			t.Fatalf("NewDownloader: %v", err)
+		}
+
+		if err := d.DownloadDirectory(ctx, &DownloadDirectoryInput{
+			Bucket:         tb.bucket,
+			LocalDirectory: localDir,
+			OnObjectDownload: func(got *DownloadOutput) {
+				callbackMu.Lock()
+				numCallbacks++
+				if got.Err != nil {
+					numErrors++
+				}
+				callbackMu.Unlock()
+
+				if slices.Contains(illegalObjs, got.Object) {
+					if got.Err == nil {
+						t.Errorf("Expected error but got nil for object %v: %v", got.Object, got.Err)
+					}
+					return
+				}
+				if got.Err != nil {
+					t.Errorf("result.Err: %v", got.Err)
+				}
+
+				if got, want := got.Attrs.Size, tb.objectSizes[got.Object]; want != got {
+					t.Errorf("expected object size %d, got %d", want, got)
+				}
+
+				path := filepath.Join(localDir, got.Object)
+
+				f, err := os.Open(path)
+				if err != nil {
+					t.Errorf("os.Open(%q): %v", path, err)
+				}
+				defer f.Close()
+
+				b := bytes.NewBuffer(make([]byte, 0, got.Attrs.Size))
+				if _, err := io.Copy(b, f); err != nil {
+					t.Errorf("io.Copy: %v", err)
+				}
+
+				if wantCRC, gotCRC := tb.contentHashes[got.Object], crc32c(b.Bytes()); gotCRC != wantCRC {
```

Comment:
I think this check shouldn't be necessary? The storage library code should check the CRC implicitly. I think you can just skip all of L514-L534.